lib/beacon/config.ex

```diff
@@ -2,81 +2,247 @@ defmodule Beacon.Config do
+              load_template: @default_load_template,
+              render_template: @default_render_template,
+              publish_page: [],
+              create_page: []
+            ]
+
+  @type option ::
+          {:site, Beacon.Types.Site.t()}
+          | {:data_source, data_source()}
+          | {:authorization_source, authorization_source()}
+          | {:css_compiler, css_compiler()}
+          | {:tailwind_config, tailwind_config()}
+          | {:live_socket_path, live_socket_path()}
+          | {:safe_code_check, safe_code_check()}
+          | {:template_formats, template_formats()}
+          | {:lifecycle, lifecycle()}
 
   @doc """
-  Build a new `%Beacon.Config{}` instance, used to hold all the configuration for each site.
+  Build a new `%Beacon.Config{}` instance to hold the entire configuration for each site.
+
+  ## Options
+
+    * `:site` - `t:Beacon.Types.Site.t/0` (required)
+
+    * `:data_source` - `t:data_source/0` (optional)
+
+    * `:authorization_source` - `t:authorization_source/0` (optional).
+    Note this config can't be `nil`. Defaults to `Beacon.Authorization.DefaultPolicy`.
+
+    * `css_compiler` - `t:css_compiler/0` (optional).
+   Defaults to `Beacon.TailwindCompiler`.
+
+    * `:tailwind_config` - `t:tailwind_config/0` (optional).
+    Defaults to `Path.join(Application.app_dir(:beacon, "priv"), "tailwind.config.js.eex")`.
+
+    * `:live_socket_path` - `t:live_socket_path/0` (optional).
+    Defaults to `"/live"`.
+
+    * `:safe_code_check` - `t:safe_code_check/0` (optional).
+    Defaults to `false`.
+
+    * `:template_formats` - `t:template_formats/0` (optional).
+    Defaults to:
+
+          [
+            {:heex, "HEEx (HTML)"},
+            {:markdown, "Markdown (GitHub Flavored version)"}
+          ]
+
+    Note that the default config is merged with your config.
+
+    * `lifecycle` - `t:lifecycle/0` (optional).
+    Note that the default config is merged with your config.
+
+  ## Example
+
+      Beacon.Config.new(
+        site: :my_site,
+        data_source: MyApp.SiteDataSource,
+        authorization_source: MyApp.SiteAuthnPolicy,
+        tailwind_config: Path.join(Application.app_dir(:my_app, "priv"), "tailwind.config.js.eex"),
+        template_formats: [
+          {:custom_format, "My Custom Format"}
+        ],
+        lifecycle: [
+          load_template: [
+            {:custom_format,
+             [
+               validate: fn template, _metadata -> MyEngine.validate(template) end
+             ]}
+          ],
+          render_template: [
+            {:custom_format",
+             [
+               assigns: fn template, %{assigns: assigns} -> MyEngine.parse_to_html(template, assigns) end,
+               compile: &Beacon.Template.HEEx.compile/2
+               eval: &Beacon.Template.HEEx.eval_ast/2
+             ]}
+          ],
+          publish_page: [
+            notify_admin: fn page -> {:cont, MyApp.send_email(page) end
+          ]
+        ]
```

Comment:
This API can be considered done but I'm open to suggestions. Initially I thought about a API to allow users adding or removing steps from HEEx or Markdown formats so they could add custom steps to transform them. But that proved to be more complex than just letting the user define the entire lifecycle (all steps to load and render a template in a given format) with the possibility of reusing steps like I've done for Markdown that's reusing HEEx steps.

And the reasons to leverage the config keyword instead of functions are to keep all config in the same place, it's easier to inspect and debug and the most important is that we need the `:load_template` steps as early as possible to boot the app so it has to be available when the site is registered.